include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,1303 @@
+  // for inferring object info:
+  //  object counting, object initial flag, <unused>
+  using obj_info_env_t =
+      ikos::separate_domain<variable_t, region_domain_impl::region_info>;
+
+  // Region representative map
+  // Map regions to object's region representative
+  using flds_rep_map_t = std::unordered_map<variable_t, variable_t>;
+  /**------------------ End type definitions ------------------**/
+
+  /**------------------ Begin field definitions ------------------**/
+  // a special symbol to represent bottom state of the object domain
+  // object domain is bottom if this flag is set or all subdomains are bottom.
+  bool m_is_bottom;
+
+  // The abstract state definition:
+  // Base domain:
+  // Domain for register, singleton object
+  base_abstract_domain_t m_base_dom;
+
+  // A map from each region variable to corresponding object domain
+  odi_map_t m_odi_map;
+
+  // Map region variables to a tuple of (Count,Init,Type)
+  //
+  // Count: count how many allocations are owned by an abstract object.
+  // Each call to ref_make models a new allocation.
+  // This allows us to decide when only if one address per object
+  // (i.e., singleton object).
+  //
+  // Init: unused.
+  //
+  // Type: unused.
+  obj_info_env_t m_obj_info_env;
+
+  // Map region variables to a representative region for an abstract object
+  // To determine the representative region,
+  // now is using the first region variable passed by the intrinsic method
+  flds_rep_map_t m_flds_rep_map;
+  /**------------------ End field definitions ------------------**/
+
+  /**------------------ Begin helper method definitions ------------------**/
+  // Constructor Definition
+  object_domain(base_abstract_domain_t &&base_dom, odi_map_t &&odi_map,
+                obj_info_env_t &&obj_info_env, flds_rep_map_t &&flds_rep_map)
+      : m_is_bottom(base_dom.is_bottom()), m_base_dom(base_dom),
+        m_odi_map(odi_map), m_obj_info_env(obj_info_env),
+        m_flds_rep_map(flds_rep_map) {}
+
+  // Self join
+  // Perform *this = join(*this, right)
+  void self_join(const object_domain_t &right) {
+
+    // 1. join the parts of each state that do not require common renaming
+    obj_info_env_t out_obj_info_env(m_obj_info_env | right.m_obj_info_env);
+
+    // 2. join the subdomain for base and objects
+    // 2.1 join the base domain
+    base_abstract_domain_t out_base_dom =
+        base_abstract_domain_t(m_base_dom | right.m_base_dom);
+
+    // 2.2 join the objects' domain over odi map
+    // In the real join operation,
+    // the following join operates on two odi maps containg the same keys
+    // even if it can performed without such assumption.
+    odi_map_t out_odi_map(m_odi_map | right.m_odi_map);
+
+    // The following map update is not necessary because m_flds_rep_map
+    // is fixed after intrinsic calls.
+    // Here, in case to keep operation sound.
+    flds_rep_map_t out_flds_rep_map;
```

Comment:
this is wrong. `flds_rep_map` must be a shared map between all values, not a copy!